### ISSUE TITLE
Clip for images 1 pixel out but works for rect objects

### ISSUE THREAD

**Issue by kylekirkby** [fabricjs/fabric.js#6325]
I've come across a strange bug when using clipping. I've got a rect object that I'm clipping my objects to. When scaling a rect object, to larger than the clipping rect, the object covers the clipping rect fully. But when adding an image and doing the same their is a border showing the rect beneath. So puzzling! @asturur do you have any ideas what could cause this? 

![Peek 2020-05-09 11-00](https://user-images.githubusercontent.com/4564433/81470744-5b62e400-91e4-11ea-8490-27a91aa860bf.gif)

Best regards,

Kyle

**Comment by asturur** [comment]
@kylekirkby please you have to link to a working example in one of the classic coding platforms. ( codepen, jsfiddle... )

**Comment by kylekirkby** [comment]
Hi @asturur,

Thanks for your reply! Apologies for not providing an example initially.

https://codepen.io/thegod/pen/zYvjPbq

You can see the issue I'm facing in this example.

Best regards,

Kyle

**Comment by asturur** [comment]
i cannot actually see the red border in the code pen, with and without retina scaling.
I tried to scale the red rectangle but i cannot get it.

Is the red border visible for you on the codepen?

A couple of suggestions, unrelated to the problem, on a macbook pro 2015 that codepen is very slow. You are using a canvas that is 4000x4000 with my retina scaling is 8000x8000 and with the clipPath on the canvas there is a double drawing each frame, resulting in sluggish performance.

If you want to create an artboard and a bleed border on the canvas there are maybe less expensive solutions. I personally draw an overlay or a custom function in the after:render event in order to mask away what i do not want from the canvas. It works well just for rectangular areas like this.

**Comment by kylekirkby** [comment]
Hi @asturur,

Yes the border is visible for me on the codepen and is also visible when saving to an image in my nodejs lambda function. :/

How do I draw and overlay or a custom function in the after:render event? That sounds very interesting would you mind sharing and example if you would be so kind :) 

The reason I’ve created a 4000 pixel canvas is so that objects can be dragged off of the template area whilst users are positioning and designing their templates.

**Comment by asturur** [comment]
that is fine but you will obtain faster performances using the fabricjs pan and zoom rather than working with a big canvas. Unless you are working on a 4k screen at full screen no borders, there little reason for a 4k canvas before export

**Comment by kylekirkby** [comment]
I see what you mean and have noticed that the performance is pretty bad right now! Will I be able to drag objects off of the canvas if i switch to a canvas that is the size of the template and then zoom out a bit first?

**Comment by kylekirkby** [comment]
![border-bug](https://user-images.githubusercontent.com/4564433/81533017-56757000-935d-11ea-90e4-dad8ee7e62fe.gif)

**Comment by asturur** [comment]
yes you will be able to drag them outside of the canvas area, the drag/move events have the handler attached to the window, so you continue to drag them away even if you get out of the canvas HTML element.

I can make an example of this, i need time because i also promised another example for soomething else.

**Comment by kylekirkby** [comment]
Sounds great! Thank you @asturur !

**Comment by kylekirkby** [comment]
Hi @asturur,

Is there no way I can fix this 1 pixel border bug that is shown in the codepen above? I've tried using the zoom and pan features but still need to clip the objects to the canvas since when dragging objects off the canvas as you can still see them. Unless I could create a clipPath the size of the canvas?

**Comment by kylekirkby** [comment]
Hi @asturur,

Any ideas what this 1px transparent border bug could be caused by? Just noticed that there is something strange going on every other pixel it seems... see below

![Peek 2020-05-17 12-43](https://user-images.githubusercontent.com/4564433/82143481-3998db00-983c-11ea-8c94-bba29ebddba0.gif)

**Comment by kylekirkby** [comment]
@asturur I'm happy to try and fix this myself if you point me to where the logic for clip paths are implemented in the source code

**Comment by asturur** [comment]
the transparent pixel is some half positioning probably.
I'm clearing out a long list of emails an i m trying to make some docs today, including that example we discussed in this issue

**Comment by kylekirkby** [comment]
> the transparent pixel is some half positioning probably.
> I'm clearing out a long list of emails an i m trying to make some docs today, including that example we discussed in this issue

Ok thank you @asturur - my world around currently is to add 2px on the width and height and then grab the image. The extra 1px around the edges mean images are saved perfectly.

**Comment by kylekirkby** [comment]
Would your example make objects not visible outside the canvas but selectable so users can still drag them back on to the canvas? I couldn’t see a way to clip the objects without a clip path ? Does the zoom and pan do this automatically ?

**Comment by juzhiyuan** [comment]
![image](https://user-images.githubusercontent.com/2106987/95214417-76349580-0822-11eb-8a5f-01ae8faaeb86.png)

How about this one?

**Comment by asturur** [comment]
yeah something like that

**Comment by juzhiyuan** [comment]
I implement that by using the following method, use a rectangle as `canvas. setBackgroundImage`'s parameter, and set a transparent canvas background.

https://github.com/juzhiyuan/blog/commit/7ebcdfb216d8118902de9dcef4599ae679b5c592#diff-85a0da20aefb1db3ade95f0a60a2f7ebR24-R38

BTW, I'm searching for how to Drag that rectangle with objects when pressing the space bar, so it could move as Photoshop does 🤔  @asturur Could you please give some tips?

**Comment by juzhiyuan** [comment]
Just found http://fabricjs.com/fabric-intro-part-5 this one, and it seems work by using viewportTransform.

**Comment by asturur** [comment]
yes exactly, viewportTransform is the result of zoom and pan.

**Comment by asturur** [comment]
Do you want to make a tutorial on this bleeding canvas? lot of people ask for it. Is enough you replicate the example in a codepen like the custom controls demo

**Comment by juzhiyuan** [comment]
> Do you want to make a tutorial on this bleeding canvas? lot of people ask for it. Is enough you replicate the example in a codepen like the custom controls demo

Ya, I would have a try on this when I free my hands 😂  btw, others could take a look at https://stackoverflow.com/a/44454016 first.

**Comment by thobn24h** [comment]
@juzhiyuan : that means, currently fabricjs doesn't support this mode for the clip path of object ? We must implement some further logic to make same effect?

**Comment by pushuo** [comment]
> ![image](https://user-images.githubusercontent.com/2106987/95214417-76349580-0822-11eb-8a5f-01ae8faaeb86.png)
> 
> How about this one?

How do I add multiple buttons to the object?

**Comment by juzhiyuan** [comment]
buttons?

**Comment by asturur** [comment]
@pushuo you can do that wiith the control API if you are willing to, or  you can do that with HTML, depends on the effect you want.
